### PR TITLE
[TH2-1672] Extract router configurations to separate files. Refactor configurations.

### PR DESCRIPTION
**Pull request by Ilya-Loginov-Exactpro**

### CODE REVIEW

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 1]
```diff
@@ -498,4 +493,228 @@ private static Path calculatePath(String path, String configsPath, String fileNa
     private static Path calculatePath(CommandLine cmd, Option option, String configs, String fileName) {
         return calculatePath(cmd.getOptionValue(option.getLongOpt()), configs, fileName);
     }
+
+    private static Path calculatePath(CommandLine cmd, Option current, Option deprecated, String configs, String fileName) {
+        return calculatePath(ObjectUtils.defaultIfNull(cmd.getOptionValue(current.getLongOpt()), cmd.getOptionValue(deprecated.getLongOpt())), configs, fileName);
+    }
+
+    public static class Settings {
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:501]: Please extract to a separate class

File: src/main/java/com/exactpro/th2/common/schema/grpc/router/AbstractGrpcRouter.java [hunk 2]
```diff
@@ -45,11 +51,28 @@
...
+    protected GrpcConfiguration configuration;
+    protected EventLoopGroup eventLoop;
 
     @Override
-    public void init(GrpcRouterConfiguration configuration) {
+    public void init(GrpcConfiguration configuration) {
+        if (this.configuration != null && eventLoop != null) {
+            throw new IllegalStateException("Grpc router already init");
+        }
+
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/common/schema/grpc/router/AbstractGrpcRouter.java:62]: Extract to separate method

File: src/main/java/com/exactpro/th2/common/schema/grpc/router/AbstractGrpcRouter.java [hunk 3]
```diff
@@ -45,11 +51,28 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(AbstractGrpcRouter.class);
     protected List<Server> servers = new ArrayList<>();
-    protected GrpcRouterConfiguration configuration;
+    protected GrpcConfiguration configuration;
+    protected EventLoopGroup eventLoop;
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/common/schema/grpc/router/AbstractGrpcRouter.java:55]: We should close the eventLoop

File: src/main/java/com/exactpro/th2/common/schema/grpc/router/GrpcRouter.java [hunk 4]
```diff
@@ -31,7 +29,10 @@
     /**
      * Initialization router
      */
-    void init(GrpcRouterConfiguration configuration);
+    @Deprecated(since = "3.9.0", forRemoval = true)
+    void init(GrpcConfiguration configuration);
+
+    void init(@NotNull GrpcConfiguration configuration, @NotNull GrpcRouterConfiguration routerConfiguration);
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/common/schema/grpc/router/GrpcRouter.java:35]: Maybe switch to constructor from init method

Comment (Ilya-Loginov-Exactpro) [reply src/main/java/com/exactpro/th2/common/schema/grpc/router/GrpcRouter.java:35]: If we will do it, we should find a special constructor for the class. We can not be sure the constructor exists, but interface requires the developer to implement `init` method.

File: src/main/java/com/exactpro/th2/common/schema/grpc/router/AbstractGrpcRouter.java [hunk 5]
```diff
@@ -120,5 +122,14 @@ public void close() {
                 LOGGER.error("Failed to shutdown server: {}", server, e);
             }
         }
+
+
+        LOGGER.info("Shutting down event loop");
+        try {
+            eventLoop.shutdownGracefully().await(SERVER_SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/common/schema/grpc/router/AbstractGrpcRouter.java:129]: Check result of shutdown gracefully operation

**Review by Nikita-Smirnov-Exactpro — COMMENTED** [review]
Please update README

File: src/main/kotlin/com/exactpro/th2/common/schema/configuration/ConfigurationManager.kt [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.concurrent.ConcurrentHashMap
+import java.util.function.Supplier
+
+class ConfigurationManager(private val configurationPath: Map<Class<*>, Path>) {
+
+    companion object {
+        private val LOGGER: Logger = LoggerFactory.getLogger(ConfigurationManager::class.java)
+    }
+
+    private val configurations: MutableMap<Class<*>, Any?> = ConcurrentHashMap()
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/configuration/ConfigurationManager.kt:34]: Why nullable value?

File: src/main/kotlin/com/exactpro/th2/common/schema/cradle/CradleConfiguration.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    var password: String? = null,
+    var cradleInstanceName: String? = null
+) : Configuration()
+
+data class CradleNonConfidentialConfiguration(
+    var timeout: Long = CassandraStorageSettings.DEFAULT_TIMEOUT,
+    var pageSize: Int = 5000,
+    var cradleMaxEventBatchSize: Long = CassandraStorageSettings.DEFAULT_MAX_EVENT_BATCH_SIZE,
+    var cradleMaxMessageBatchSize: Long = CassandraStorageSettings.DEFAULT_MAX_MESSAGE_BATCH_SIZE
+) : Configuration()
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/cradle/CradleConfiguration.kt:69]: Why all these parameters are `var`? Are we going to change them anywhere?

File: src/main/kotlin/com/exactpro/th2/common/schema/factory/FactorySettings.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    var connectionManagerSettings: Path? = null,
+    var grpc: Path? = null,
+    var routerGRPC: Path? = null,
+    var cradleConfidential: Path? = null,
+    var cradleNonConfidential: Path? = null,
+    var prometheus: Path? = null,
+    var boxConfiguration: Path? = null,
+    private var _custom: Path? = null,
+    private var _dictionariesDir: Path? = null,
+    var variables: Map<String, String> = emptyMap()) {
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/factory/FactorySettings.kt:59]: Why all parameters are `var`?

File: src/main/kotlin/com/exactpro/th2/common/schema/factory/FactorySettings.kt [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        paths[RabbitMQConfiguration::class.java] = defaultPathIfNull(rabbitMQ, CommonFactory.RABBIT_MQ_FILE_NAME)
+        paths[MessageRouterConfiguration::class.java] = defaultPathIfNull(routerMQ, CommonFactory.ROUTER_MQ_FILE_NAME)
+        paths[ConnectionManagerConfiguration::class.java] = defaultPathIfNull(connectionManagerSettings, CommonFactory.CONNECTION_MANAGER_CONF_FILE_NAME)
+        paths[GrpcConfiguration::class.java] = defaultPathIfNull(grpc, CommonFactory.GRPC_FILE_NAME)
+        paths[GrpcRouterConfiguration::class.java] = defaultPathIfNull(routerGRPC, CommonFactory.ROUTER_GRPC_FILE_NAME)
+        paths[CradleConfidentialConfiguration::class.java] = defaultPathIfNull(cradleConfidential, CommonFactory.CRADLE_CONFIDENTIAL_FILE_NAME)
+        paths[CradleNonConfidentialConfiguration::class.java] = defaultPathIfNull(cradleNonConfidential, CommonFactory.CRADLE_NON_CONFIDENTIAL_FILE_NAME)
+        paths[PrometheusConfiguration::class.java] = defaultPathIfNull(prometheus, CommonFactory.PROMETHEUS_FILE_NAME)
+        paths[BoxConfiguration::class.java] = defaultPathIfNull(boxConfiguration, CommonFactory.BOX_FILE_NAME)
+    })
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/factory/FactorySettings.kt:79]: Why the settings class creates the configuration manager? Probably, we should have the factory method in the ConfigurationManager to create the manager using FactorySettings

File: src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcConfiguration.kt [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        (minMethodRetriesTimeout + if (maxAttempts > 1) (maxMethodRetriesTimeout - minMethodRetriesTimeout) / (maxAttempts - 1) * index else 0)
+
+    override fun getMaxAttempts(): Int = maxAttempts
+    fun setMaxAttempts(maxAttempts: Int) {
+        this.maxAttempts = maxAttempts
+    }
+}
+
+data class GrpcServerConfiguration(
+    @JsonProperty(required = true) var host: String? = null,
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcConfiguration.kt:56]: Why the parameter is required and it has the default value?

File: src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcConfiguration.kt [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+data class GrpcConfiguration(
+    @JsonProperty var services: Map<String, GrpcServiceConfiguration> = emptyMap(),
+    @JsonProperty(value = "server") var serverConfiguration: GrpcServerConfiguration = GrpcServerConfiguration(),
+    @JsonProperty var retryConfiguration: GrpcRetryConfiguration = GrpcRetryConfiguration()
+) : Configuration()
+
+data class GrpcServiceConfiguration(
+    @JsonProperty(required = true) var strategy: RoutingStrategy<*>? = null,
+    @JsonProperty(value = "service-class", required = true) var serviceClass: Class<*>? = null,
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcConfiguration.kt:31]: Why the required parameters are `null`?

File: src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcRawFilterStrategy.kt [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+package com.exactpro.th2.common.schema.grpc.configuration
+
+import com.exactpro.th2.common.schema.message.configuration.FieldFilterConfiguration
+import com.exactpro.th2.common.schema.message.configuration.RouterFilter
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import com.fasterxml.jackson.annotation.JsonProperty
+
+data class GrpcRawFilterStrategy(var filters: List<GrpcRouterFilterConfiguration?> = emptyList())
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcRawFilterStrategy.kt:25]: Why do we ignore unknown properties?

File: src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcRawFilterStrategy.kt [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+data class GrpcRouterFilterConfiguration(
+    @JsonProperty(required = true) var endpoint: String? = null,
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcRawFilterStrategy.kt:27]: Why the required parameter has a default value?

File: src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcRawRobinStrategy.kt [hunk 14]
```diff
@@ -14,18 +13,8 @@
...
 
-    @JsonProperty(required = true)
-    protected List<String> endpoints;
+import com.fasterxml.jackson.annotation.JsonProperty
 
-    public List<String> getEndpoints() {
-        return endpoints;
-    }
-}
+data class GrpcRawRobinStrategy(@JsonProperty(required = true) var endpoints: List<String?> = emptyList())
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcRawRobinStrategy.kt:20]: Why the list have nullable values?

File: src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcConfiguration.kt [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    @JsonProperty(required = true) var endpoints: Map<String, GrpcEndpointConfiguration> = emptyMap()
+) : Configuration()
+
+data class GrpcEndpointConfiguration(
+    @JsonProperty(required = true) var host: String? = null,
+    @JsonProperty(required = true) var port: Int? = null,
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/grpc/configuration/GrpcConfiguration.kt:37]: Why the parameters are required and have default values? Also, why are they `null`?

File: src/main/kotlin/com/exactpro/th2/common/schema/message/configuration/MessageRouterConfiguration.kt [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+data class QueueConfiguration(
+    @JsonProperty(required = true) @JsonAlias("name", "sendKey") var routingKey: String,
+    @JsonProperty(required = true) @JsonAlias("subscribeKey") var queue: String,
+    @JsonProperty(required = true) var exchange: String,
+    @JsonProperty(required = true) @JsonAlias("labels", "tags") var attributes: List<String> = emptyList(),
+    var filters: List<MqRouterFilterConfiguration> = emptyList(),
+    @JsonProperty(value = "read", defaultValue = "true") var isReadable: Boolean = true,
+    @JsonProperty(value = "write", defaultValue = "true") var isWritable: Boolean = true
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/message/configuration/MessageRouterConfiguration.kt:41]: The `defaultValue` value does not work for constructor parameters. The Kotlin itself provides feature for defining the default value

File: src/main/kotlin/com/exactpro/th2/common/schema/message/configuration/MessageRouterConfiguration.kt [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+) : Configuration()
+
+data class MqRouterFilterConfiguration(
+    override var metadata: Map<String, FieldFilterConfiguration> = emptyMap(),
+    override var message: Map<String, FieldFilterConfiguration> = emptyMap()
+) : Configuration(), RouterFilter
+
+data class FieldFilterConfiguration(
+    var value: String? = null,
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/message/configuration/MessageRouterConfiguration.kt:50]: Why the parameter is nullable?

File: src/test/kotlin/com/exactpro/th2/common/schema/TestJsonConfiguration.kt [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+        init {
+            val module = SimpleModule()
+            module.addDeserializer(RoutingStrategy::class.java, JsonDeserializerRoutingStategy())
+            module.addSerializer(RoutingStrategy::class.java, JsonSerializerRoutingStrategy(ObjectMapper().apply { registerModule(KotlinModule()) }))
+
+            OBJECT_MAPPER.registerModule(module)
+            OBJECT_MAPPER.registerModule(KotlinModule())
+        }
+    }
```
Comment (OptimumOpium) [inline src/test/kotlin/com/exactpro/th2/common/schema/TestJsonConfiguration.kt:56]: I think we should use the same MAPPER as we do in Factory

File: src/main/kotlin/com/exactpro/th2/common/schema/message/impl/rabbitmq/configuration/RabbitMQConfiguration.kt [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+data class RabbitMQConfiguration(
+    var host: String? = null,
+    @JsonProperty("vHost") var vHost: String? = null,
+    var port: Int = 0,
+    var username: String? = null,
+    var password: String? = null,
+    @Deprecated(message = "Please use subscriber name from ConnectionManagerConfiguration")
+    var subscriberName: String? = null,
+    var exchangeName: String? = null) : Configuration()
```
Comment (OptimumOpium) [inline src/main/kotlin/com/exactpro/th2/common/schema/message/impl/rabbitmq/configuration/RabbitMQConfiguration.kt:29]: Why parameters are nullable?

**Review by OptimumOpium — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/exactpro/th2/common/schema/message/configuration/MessageRouterConfiguration.kt (hunk 17 above)
Comment (Nikita-Smirnov-Exactpro) [reply src/main/kotlin/com/exactpro/th2/common/schema/message/configuration/MessageRouterConfiguration.kt:50]: Value can be null when operation is EMPTY or NOT_EMPTY

File: src/main/kotlin/com/exactpro/th2/common/schema/cradle/CradleConfiguration.kt (hunk 7 above)
Comment (Nikita-Smirnov-Exactpro) [reply src/main/kotlin/com/exactpro/th2/common/schema/cradle/CradleConfiguration.kt:69]: For example in tests

File: README.md [hunk 20]
```diff
@@ -42,6 +42,9 @@ Then you will create an instance of imported class, by choosing one of the follo
     * --namespace - the namespace in Kubernetes to search config maps
     * --boxName - the name of the target th2 box placed in the specified Kubernetes namespace
     * --contextName - the context name to search connect parameters in Kube config
+    * --dictionaries - the mapping between a dictionary in infra schema and a dictionary type in the format: 
+      `--dictionaries <dictionary name>=<dictionary type >[ <dictionary name>=<dictionary type >]`. 
+      It can be useful when you required dictionaries to start box.
```
Comment (jean-michel-perli-sonsin) [inline README.md:47]:
```suggestion
      It can be useful when you required dictionaries to start **a specific** box. 
```

File: README.md [hunk 21]
```diff
@@ -42,6 +42,9 @@ Then you will create an instance of imported class, by choosing one of the follo
     * --namespace - the namespace in Kubernetes to search config maps
     * --boxName - the name of the target th2 box placed in the specified Kubernetes namespace
     * --contextName - the context name to search connect parameters in Kube config
+    * --dictionaries - the mapping between a dictionary in infra schema and a dictionary type in the format: 
+      `--dictionaries <dictionary name>=<dictionary type >[ <dictionary name>=<dictionary type >]`. 
+      It can be useful when you required dictionaries to start box. 
     
     Their usage is discovered further.
```
Comment (jean-michel-perli-sonsin) [inline README.md:49]:
```suggestion
    Their usage is **disclosed** further.
```

File: src/main/java/com/exactpro/th2/common/schema/grpc/router/GrpcRouter.java [hunk 22]
```diff
@@ -16,12 +15,11 @@
...
-
+import com.exactpro.th2.common.schema.grpc.configuration.GrpcConfiguration;
 import com.exactpro.th2.common.schema.grpc.configuration.GrpcRouterConfiguration;
-
 import io.grpc.BindableService;
 import io.grpc.Server;
+import org.jetbrains.annotations.NotNull;
 
 /**
  * Interface for create service for manage connections
```
Comment (jean-michel-perli-sonsin) [inline src/main/java/com/exactpro/th2/common/schema/grpc/router/GrpcRouter.java:25]:
```suggestion
 * Interface **used** for **creating services** for **managing** connections
```

File: src/main/java/com/exactpro/th2/common/schema/grpc/router/GrpcRouter.java [hunk 23]
```diff
@@ -31,7 +29,10 @@
...
      * Initialization router
      */
-    void init(GrpcRouterConfiguration configuration);
+    @Deprecated(since = "3.9.0", forRemoval = true)
+    void init(GrpcConfiguration configuration);
+
+    void init(@NotNull GrpcConfiguration configuration, @NotNull GrpcRouterConfiguration routerConfiguration);
 
     /**
      * Create grpc service for send message to grpc servers
```
Comment (jean-michel-perli-sonsin) [inline src/main/java/com/exactpro/th2/common/schema/grpc/router/GrpcRouter.java:38]:
```suggestion
     * Create grpc service for **sending messages** to grpc servers
```

**Review by jean-michel-perli-sonsin — CHANGES_REQUESTED** [review]
Hello, please see if you agree with the suggestions proposed below:
README.md
Line 53/56 - Create factory with a namespace in Kubernetes, the name of the target th2 box from Kubernetes and the **context name** to choose **a specific** context from Kube config: 

src/main/kotlin/com/exactpro/th2/common/metrics/PrometheusConfiguration.kt
Line 2 - Need to update the copyright to 2020-2021 (like in the other files)

The other suggestions proposed were written at the file itself.
Please let me know in case of any question. Thanks!

**Review by jean-michel-perli-sonsin — DISMISSED** [review]
All proposed changes were corrected accordingly. Thanks!

File: src/main/kotlin/com/exactpro/th2/common/schema/cradle/CradleConfiguration.kt (hunk 7 above)
Comment (OptimumOpium) [reply src/main/kotlin/com/exactpro/th2/common/schema/cradle/CradleConfiguration.kt:69]: Why do you need to create the instance in the test and then change its properties?

**Review by OptimumOpium — APPROVED** [review]

**Pull request merged**